# Description   

Let's practice and become familiar with regression.

Level 1

- Exercise 1

Create at least three different regression models to try to predict as well as possible the flight delay (ArrDelay) of DelayedFlights.csv.

- Exercise 2

Compare them on the basis of MSE and R2 .

- Exercise 3

Train them using the different parameters they allow.

- Exercise 4

Compare their performance using the test/train approach or using all data (internal validation).

Level 2

- Exercise 5

Perform some variable engineering process to improve your prediction.

Level 3

- Exercise 6

Do not use the DepDelay variable when making predictions.

# Level 1  

## - Exercise 1

Create at least three different regression models to try to predict as well as possible the flight delay (ArrDelay) of DelayedFlights.csv.

In [ ]:
# Import libraries
import pandas as pd
import numpy as np
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import cross_val_score

In [ ]:
# settings to display all columns (default is 20, now is None (all))
pd.set_option("display.max_columns", None)

In [ ]:
# Import cleaned and sampled train an test dataset from previous Task.
df_train = pd.read_csv('..\data\DelayedFlights_train.csv')
df_test  = pd.read_csv('..\data\DelayedFlights_test.csv')

In [77]:
# Let's explore the dataset
display(df_train.head())
display(df_train.shape)
display(df_test.head())
display(df_test.shape)

DayOfWeek   DepTime  CRSDepTime   ArrTime  CRSArrTime  ActualElapsedTime  \
0          2  1.396108    1.553827  1.265258    1.312513          -0.781188   
1          2 -0.699602   -0.997079 -0.540408   -0.920752          -0.905020   
2          4  0.049503    0.125887  0.575125    0.600955           0.402102   
3          4 -0.445441   -0.394224  0.222755    0.198024           0.801118   
4          5  0.004914   -0.082158  0.184414   -0.012015          -0.340894   

   CRSElapsedTime   AirTime  ArrDelay  DepDelay  Distance    TaxiIn   TaxiOut  \
0       -0.790736 -0.818653 -0.423085 -0.453348 -0.683070 -0.341142  0.130433   
1       -0.832512 -0.804233  0.577369  0.739595 -0.783172 -0.154948 -0.654312   
2        0.490397  0.551251 -0.536344 -0.453348  0.708000 -0.713531 -0.368950   
3        0.936008  0.868492 -0.630727 -0.493787  0.918559 -0.341142 -0.012248   
4       -0.317274 -0.227431  0.067703  0.112794 -0.334439  0.217440 -0.725652   

         Date  UniqueCarrier_9E  UniqueCarrier_AA  UniqueCarrier_AQ  \
0  2008-07-29                 0                 0                 0   
1  2008-10-07                 0                 0                 0   
2  2008-06-26                 0                 0                 0   
3  2008-04-10                 0                 0                 0   
4  2008-01-11                 0                 0                 0   

   UniqueCarrier_AS  UniqueCarrier_B6  UniqueCarrier_CO  UniqueCarrier_DL  \
0                 0                 0                 0                 0   
1                 0                 0                 1                 0   
2                 0                 0                 0                 0   
3                 0                 0                 0                 0   
4                 0                 0                 0                 0   

   UniqueCarrier_EV  UniqueCarrier_F9  UniqueCarrier_FL  UniqueCarrier_HA  \
0                 0                 0                 0                 0   
1                 0                 0                 0                 0   
2                 0                 0                 0                 0   
3                 0                 0                 0                 0   
4                 0                 0                 1                 0   

   UniqueCarrier_MQ  UniqueCarrier_NW  UniqueCarrier_OH  UniqueCarrier_OO  \
0                 0                 1                 0                 0   
1                 0                 0                 0                 0   
2                 0                 0                 0                 0   
3                 0                 0                 0                 0   
4                 0                 0                 0                 0   

   UniqueCarrier_UA  UniqueCarrier_US  UniqueCarrier_WN  UniqueCarrier_XE  \
0                 0                 0                 0                 0   
1                 0                 0                 0                 0   
2                 0                 0                 1                 0   
3                 0                 1                 0                 0   
4                 0                 0                 0                 0   

   UniqueCarrier_YV  
0                 0  
1                 0  
2                 0  
3                 0  
4                 0

(12919, 34)

DayOfWeek   DepTime  CRSDepTime   ArrTime  CRSArrTime  ActualElapsedTime  \
0          6 -1.823944   -1.944788 -1.260533   -1.568576           0.189117   
1          1  0.215545    0.130093  0.194082   -0.020472          -0.092822   
2          2 -0.900783   -0.990819 -0.112347   -0.211244           0.344183   
3          7  0.756863    0.893268  1.347805    1.496933           0.527443   
4          1  0.015639   -0.311117  0.343604   -0.059942          -0.515730   

   CRSElapsedTime   AirTime  ArrDelay  DepDelay  Distance    TaxiIn   TaxiOut  \
0        0.250020  0.109294 -0.582996 -0.546662  0.005498  1.589749 -0.155199   
1       -0.091189 -0.052790  0.177841  0.194912 -0.116385  0.239096 -0.293704   
2        0.520144  0.462933 -0.528651 -0.331996  0.729733 -0.146805 -0.432210   
3        0.363756  0.477668 -0.311269 -0.566177  0.576054 -0.146805  0.398824   
4       -0.560351 -0.465368  1.246637  1.287758 -0.409612  1.203848 -0.778474   

         Date  UniqueCarrier_9E  UniqueCarrier_AA  UniqueCarrier_AQ  \
0  2008-05-10                 0                 0                 0   
1  2008-12-01                 0                 0                 0   
2  2008-12-23                 0                 1                 0   
3  2008-08-17                 0                 0                 0   
4  2008-01-21                 0                 0                 0   

   UniqueCarrier_AS  UniqueCarrier_B6  UniqueCarrier_CO  UniqueCarrier_DL  \
0                 0                 0                 0                 1   
1                 0                 0                 0                 0   
2                 0                 0                 0                 0   
3                 0                 0                 0                 0   
4                 0                 0                 0                 0   

   UniqueCarrier_EV  UniqueCarrier_F9  UniqueCarrier_FL  UniqueCarrier_HA  \
0                 0                 0                 0                 0   
1                 0                 0                 1                 0   
2                 0                 0                 0                 0   
3                 0                 1                 0                 0   
4                 0                 0                 0                 0   

   UniqueCarrier_MQ  UniqueCarrier_NW  UniqueCarrier_OH  UniqueCarrier_OO  \
0                 0                 0                 0                 0   
1                 0                 0                 0                 0   
2                 0                 0                 0                 0   
3                 0                 0                 0                 0   
4                 0                 0                 0                 1   

   UniqueCarrier_UA  UniqueCarrier_US  UniqueCarrier_WN  UniqueCarrier_XE  \
0                 0                 0                 0                 0   
1                 0                 0                 0                 0   
2                 0                 0                 0                 0   
3                 0                 0                 0                 0   
4                 0                 0                 0                 0   

   UniqueCarrier_YV  
0                 0  
1                 0  
2                 0  
3                 0  
4                 0

(6364, 34)

### Explanation  
The train / test sample is imported from previous Task (S09T01).  
This train / test sample is 1% of the original dataset, parted 33% test and 66% train. 

In [ ]:
# Let's delete the first column
df_train = df_train.drop(columns='Unnamed: 0')
df_test  = df_test.drop(columns='Unnamed: 0')

## 1st model: Linear regression between DepDelay and ArrDelay.

In [ ]:
# Our y or Target is ArrDelay:
y_train = df_train.ArrDelay
print(type(y_train))
y_train = y_train.array # let's convert Pd series column to Numpy Array as needed for the model.
print(type(y_train))
y_test  = df_test.ArrDelay
y_test  = y_test.array # let's convert Pd series column to Numpy Array as needed for the model.

In [ ]:
# Our x easily could be DepDelay (as logical)
x_train = df_train.DepDelay
print("Shape before reshape: ",x_train.shape)
x_train = x_train.array.reshape(-1, 1)
print("Shape after reshape: ",x_train.shape)
x_test  = df_test.DepDelay
x_test  = x_test.array.reshape(-1, 1)

In [ ]:
# Fit the model ( calculate b0 and b1 for the model y = b0 + b1x )
model1 = LinearRegression().fit(x_train,y_train)

In [33]:
# Calculate R2 to see the accuracy of the model with train data.
r_sq_train = model1.score(x_train,y_train)
print('coefficient of determination with train data: %.3f' %r_sq_train)

coefficient of determination with train data: 0.895


In [34]:
# Calculate R2 to see the accuracy of the model with test data.
r_sq_test = model1.score(x_test,y_test)
print('coefficient of determination with test data: %.3f' %r_sq_test)

coefficient of determination with test data: 0.902


In [35]:
# Let's calculate the prediction of the model
y_pred1 = model1.predict(x_train)
y_pred1.shape

(12919,)

### Observation 1:  
As we expected, the accuracy of the model is very high (0.895 with train data and 0.902 with test data).  
The dependency between Arrival Delay and Departure Delay is quite obvious.

## 2nd model: Multiple Linear Regression

In [36]:
# Our Y or Target is ArrDelay:
y_train = df_train.ArrDelay
y_train = y_train.array # Convert pandas series to numpy array
type(y_train)
y_test  = df_test.ArrDelay.array


In [37]:
# Our X now is going to be all the columns in df_train except ArrDelay and Date (also the OHE of Airline)
X_train = df_train.drop(columns=["ArrDelay","Date"])
feature_list = list(X_train.columns) # Saving feature names for later use
X_train = X_train.to_numpy() # Convert dataframe to array
type(X_train)
X_test = df_test.drop(columns=["ArrDelay","Date"]).to_numpy()

In [38]:
# Fit the model ( calculate b0...bn for the multiple lineal model y = b0 + b1x1 +... + bnxn)
model2 = LinearRegression().fit(X_train,y_train)

In [39]:
# Calculate R2 to see the accuracy of the model with train data.
r_sq_train2 = model2.score(X_train,y_train)
print('coefficient of determination with train data: %.3f' %r_sq_train)

coefficient of determination with train data: 0.895


In [40]:
# Calculate R2 to see the accuracy of the model with test data.
r_sq_test2 = model2.score(X_test,y_test)
print('coefficient of determination with test data: %.3f' %r_sq_test)

coefficient of determination with test data: 0.902


In [41]:
# Applying k-Fold Cross Validation (CV) with train data
accuracies2a = cross_val_score(estimator = model2, X=X_train, y=y_train , cv = 10)
print("Multiple Linear Regression:\n Accuracy with train data:", accuracies2a.mean(), "+/-", accuracies2a.std(),"\n")

Multiple Linear Regression:
 Accuracy with train data: 1.0 +/- 0.0 



In [42]:
# Applying k-Fold Cross Validation (CV) with test data
accuracies2b = cross_val_score(estimator = model2, X=X_test, y=y_test , cv = 10)
print("Multiple Linear Regression:\n Accuracy with test data:", accuracies2b.mean(), "+/-", accuracies2b.std(),"\n")

Multiple Linear Regression:
 Accuracy with test data: 1.0 +/- 0.0 



In [43]:
# Let's calculate the prediction of the model
y_pred2 = model2.predict(X_train)
y_pred2.shape

(12919,)

### Observation 2:  
The accuracy of the model is the maximum (1.0) with Train data.    
That is a perfect fit, meaning that all the data is explained with the model. Like the target is a dependent variable of the attributes.  
That's normal, because is our $R^2$ with the train set. 

Later, we test it against the test set and $R^2$ and is also a perfect fit.  
This is not normal in the real world, but we have a very dependant variable DepDelay, that can explain this behavior and also the accuracy of Cross Validation.

## 3rd model: Random Forest Regression

In [44]:
# Let's create the model of Random Forest
# Import the model we are using
from sklearn.ensemble import RandomForestRegressor
# Instantiate model with 100 decision trees
model3 = RandomForestRegressor(n_estimators = 10, random_state = 42)
# Train the model on training data
model3.fit(X_train, y_train)

RandomForestRegressor(n_estimators=10, random_state=42)

In [45]:
# Calculate R2 to see the accuracy of the model with train data.
r_sq_train3 = model3.score(X_train,y_train)
print('coefficient of determination with train data: %.3f' %r_sq_train2)

coefficient of determination with train data: 1.000


In [46]:
# Calculate R2 to see the accuracy of the model with test data.
r_sq_test3 = model3.score(X_test,y_test)
print('coefficient of determination with test data: %.3f' %r_sq_test2)

coefficient of determination with test data: 1.000


In [47]:
# Applying k-Fold Cross Validation (CV)
accuracies3 = cross_val_score(estimator = model3, X=X_train, y=y_train , cv = 10)
print("RandomForestRegressor:\n Accuracy:", accuracies3.mean(), "+/-", accuracies3.std(),"\n")

RandomForestRegressor:
 Accuracy: 0.9565624734757234 +/- 0.007685769658961897 



In [48]:
# Applying k-Fold Cross Validation (CV) with train data
accuracies3a = cross_val_score(estimator = model3, X=X_train, y=y_train , cv = 10)
print("RandomForest Regression:\n Accuracy:", accuracies3a.mean(), "+/-", accuracies3a.std(),"\n")

RandomForest Regression:
 Accuracy: 0.9565624734757234 +/- 0.007685769658961897 



In [49]:
# Applying k-Fold Cross Validation (CV) with test data
accuracies3b = cross_val_score(estimator = model3, X=X_test, y=y_test , cv = 10)
print("RandomForest Regression:\n Accuracy:", accuracies3b.mean(), "+/-", accuracies3b.std(),"\n")

RandomForest Regression:
 Accuracy: 0.9540780342184462 +/- 0.006585823503450868 



In [50]:
# Let's calculate the prediction of the model
y_pred3 = model3.predict(X_train)
y_pred3.shape

(12919,)

### Observation 3.
* Here also the R^2 and the accuracy of the Cross Validation are very high (~0.95).
* The explanation is the same than in the previous model, the variable DepDelay is strongly dependent with the target.
* On the last exercise, we will take out this variable and rerun the models.

## 4th model: Neural Network Regression

In [51]:
# Import modules
from sklearn.neural_network import MLPRegressor

In [52]:
# Create model
model4 = MLPRegressor(activation='relu',solver='adam',random_state=1, max_iter=500)
model4.fit(X_train,y_train)

MLPRegressor(max_iter=500, random_state=1)

In [53]:
# Calculate R2 to see the accuracy of the model with train data.
r_sq_train4 = model4.score(X_train,y_train)
print('coefficient of determination with train data: %.3f' %r_sq_train4)

coefficient of determination with train data: 0.999


In [54]:
# Calculate R2 to see the accuracy of the model with test data.
r_sq_test4 = model4.score(X_test,y_test)
print('coefficient of determination with test data: %.3f' %r_sq_test4)

coefficient of determination with test data: 0.998


In [55]:
# Applying k-Fold Cross Validation (CV) with train data
accuracies4a = cross_val_score(estimator = model4, X=X_train, y=y_train , cv = 10)
print("Neural Network Regression:\n Accuracy:", accuracies4a.mean(), "+/-", accuracies4a.std(),"\n")

Neural Network Regression:
 Accuracy: 0.9987745392955016 +/- 0.00011903702763198072 



In [56]:
# Applying k-Fold Cross Validation (CV) with test data
accuracies4b = cross_val_score(estimator = model4, X=X_test, y=y_test , cv = 10)
print("Neural Network Regression:\n Accuracy:", accuracies4b.mean(), "+/-", accuracies4b.std(),"\n")

Neural Network Regression:
 Accuracy: 0.9978135426959385 +/- 0.00028245018115084715 



In [57]:
# Let's calculate the prediction of the model
y_pred4 = model4.predict(X_train)
y_pred4.shape

(12919,)

### Observation 4.
* Here also the R^2 and the accuracy of the Cross Validation are very high (~0.99).
* The explanation is the same than in the previous model, the variable DepDelay is strongly dependent with the target.
* On the last exercise, we will take out this variable and rerun the models.

## Exercise 2 - Compare them on the basis of MSE and R2 .
 

R^2 has been calculated before, we copy here the results:


In [58]:
print("--------------Linear Regression Model------------")
print('R^2 - coefficient of determination with train data: %.3f' %r_sq_train)
print('R^2 - coefficient of determination with test data: %.3f' %r_sq_test)
print("-------Multiple Linear Regression Model------------")
print('R^2 - coefficient of determination with train data: %.3f' %r_sq_train2)
print('R^2 - coefficient of determination with test data: %.3f' %r_sq_test2)
print("--------------Random Forest Regression Model------------")
print('R^2 - coefficient of determination with train data: %.3f' %r_sq_train3)
print('R^2 - coefficient of determination with test data: %.3f' %r_sq_test3)
print("-------Neural Network Regression Model------------")
print('R^2 - coefficient of determination with train data: %.3f' %r_sq_train4)
print('R^2 - coefficient of determination with test data: %.3f' %r_sq_test4)

--------------Linear Regression Model------------
R^2 - coefficient of determination with train data: 0.895
R^2 - coefficient of determination with test data: 0.902
-------Multiple Linear Regression Model------------
R^2 - coefficient of determination with train data: 1.000
R^2 - coefficient of determination with test data: 1.000
--------------Random Forest Regression Model------------
R^2 - coefficient of determination with train data: 0.992
R^2 - coefficient of determination with test data: 0.965
-------Neural Network Regression Model------------
R^2 - coefficient of determination with train data: 0.999
R^2 - coefficient of determination with test data: 0.998


### Observation 5.
As seen, all R^2 are very high, maximum is Multiple Linear Regression Model, a perfect fit, all the target is explained and predicted with the model.


MSE has not been calculated, we calculate it now and print here the results:

In [59]:
from sklearn.metrics import mean_squared_error
print("--------------Linear Regression Model------------")
print('MSE - Mean Square Error with train data: %.3f' %mean_squared_error(y_true=y_train,y_pred=y_pred1))
print("-------Multiple Linear Regression Model------------")
print('MSE - Mean Square Error with train data: %.3f' %mean_squared_error(y_train,y_pred2))
print("--------Random Forest Regression Model------------")
print('MSE - Mean Square Error with train data: %.3f' %mean_squared_error(y_train,y_pred3))
print("-------Neural Network Regression Model------------")
print('MSE - Mean Square Error with train data: %.3f' %mean_squared_error(y_train,y_pred4))


--------------Linear Regression Model------------
MSE - Mean Square Error with train data: 0.105
-------Multiple Linear Regression Model------------
MSE - Mean Square Error with train data: 0.000
--------Random Forest Regression Model------------
MSE - Mean Square Error with train data: 0.008
-------Neural Network Regression Model------------
MSE - Mean Square Error with train data: 0.001


### Observation 6.
As expected from R^2 results, MSE is very low. Maximum is again Multiple Linear Regression Model, a perfect fit, MSE = 0. Let's see when we take off DepDelay.

## Exercise 3 - Train them using the different parameters they allow.


### 3.1.- Multiple Linear Regression Model


In [60]:
# This is our original model:
# Fit the model ( calculate b0...bn for the multiple lineal model y = b0 + b1x1 +... + bnxn)
model2 = LinearRegression().fit(X_train,y_train)
# Calculate R2 to see the accuracy of the model with train data.
r_sq_train2 = model2.score(X_train,y_train)
print('coefficient of determination with train data: %.3f' %r_sq_train)


coefficient of determination with train data: 0.895


In [61]:
# Let's now use different parameters: 
model2 = LinearRegression(fit_intercept=False).fit(X_train,y_train)
# Calculate R2 to see the accuracy of the model with train data.
r_sq_train2 = model2.score(X_train,y_train)
print('coefficient of determination with train data: %.3f' %r_sq_train)

# We don't change normalize= because is going to be deprecated and we have normalized before.
# copy_X will only change if we want to change X or not. By default is copied (not changed), if we use False, X will be overwritten.

coefficient of determination with train data: 0.895


In [62]:
# n_jobs: number of jobs to use for the computation. By default is None, interpreted usually as 1 CPU.
# if we use -1 all CPUs will be used in Parallel. Only effective with large datasets. Right now our 
# computational time for this model is 0.1, no improvement expected.
model2 = LinearRegression(n_jobs=-1).fit(X_train,y_train)
# Calculate R2 to see the accuracy of the model with train data.
r_sq_train2 = model2.score(X_train,y_train)
print('coefficient of determination with train data: %.3f' %r_sq_train)

coefficient of determination with train data: 0.895


#### Observation 7.  
* No improvement / change with the parameters used, speed is the same (0.1s) and R^2 is the same (0.895)

### 3.2.- Random Forest Regression Model


In [63]:
# Instantiate model with 100 decision trees
model3 = RandomForestRegressor(n_estimators = 100, random_state = 42)
# Train the model on training data
model3.fit(X_train, y_train)
# Calculate R2 to see the accuracy of the model with train data.
r_sq_train3 = model3.score(X_train,y_train)
print('coefficient of determination with train data: %.3f' %r_sq_train2)

coefficient of determination with train data: 1.000


In [64]:
# Let's change parameters, but with R2 of 1 and speed 13.4s is going to be difficult to see differences.

# The most important parameter is n_estimators, how many "trees" are in the forest. 
# By default is 100, let's change it to 10
model3 = RandomForestRegressor(n_estimators = 10, random_state = 42)
# Train the model on training data
model3.fit(X_train, y_train)
# Calculate R2 to see the accuracy of the model with train data.
r_sq_train3 = model3.score(X_train,y_train)
print('coefficient of determination with train data: %.3f' %r_sq_train2)


coefficient of determination with train data: 1.000


In [65]:
# The speed has improved to 1.5s and R2 is still 1. 
# If we change it to 1000 trees, R2 is the same, but time is 15h.

# criterion parameter also changes the function to measure the quality of a split.
# By default is squared_error, let's change it to absolute_error (poisson is not possible because there are negative values on y array)
model3 = RandomForestRegressor(n_estimators = 10, criterion='absolute_error', random_state = 42)
# Train the model on training data
model3.fit(X_train, y_train)
# Calculate R2 to see the accuracy of the model with train data.
r_sq_train3 = model3.score(X_train,y_train)
print('coefficient of determination with train data: %.3f' %r_sq_train2)



coefficient of determination with train data: 1.000


In [66]:
# There are many more parameters to change, let's just make a last change, in order to try to improve the speed, lets compute again 
# the previous model, but with using all CPUs in parallel.
model3 = RandomForestRegressor(n_estimators = 10, criterion='absolute_error', n_jobs= -1, random_state = 42)
# Train the model on training data
model3.fit(X_train, y_train)
# Calculate R2 to see the accuracy of the model with train data.
r_sq_train3 = model3.score(X_train,y_train)
print('coefficient of determination with train data: %.3f' %r_sq_train2)

coefficient of determination with train data: 1.000


#### Observation 8.  
* The speed has effectively improved with n_jobs = -1, from 46s to 13s. But CPU Performance goes to 100%!
* With R2 always 1, we are not seen difference in accuracy changing the parameters. Maybe when we take off the dependent variable these results will change.
* See comments in code, we have changed the numbers of trees, affecting to the calculation speed of the model. 
* Also criterion was changed, but with no different results. (R2 = 1)


### 3.3.- Neural Network Regression Model

In [67]:
# Create model
model4 = MLPRegressor(activation='relu',solver='adam',random_state=1, max_iter=500)
model4.fit(X_train,y_train)
# Calculate R2 to see the accuracy of the model with train data.
r_sq_train4 = model4.score(X_train,y_train)
print('coefficient of determination with train data: %.3f' %r_sq_train4)

coefficient of determination with train data: 0.999


In [73]:
# Let's increase to 5 times default number of hidden layers
model4 = MLPRegressor(hidden_layer_sizes=(500,), activation='relu',solver='adam',random_state=1, max_iter=500)
model4.fit(X_train,y_train)
# Calculate R2 to see the accuracy of the model with train data.
r_sq_train4 = model4.score(X_train,y_train)
print('coefficient of determination with train data: %.3f' %r_sq_train4)

coefficient of determination with train data: 1.000


In [74]:
# Let's change the activation to logistic sigmoid function, returns f(x) = 1 / (1 + exp(-x)). (old school)
model4 = MLPRegressor(activation='logistic',solver='adam',random_state=1, max_iter=500)
model4.fit(X_train,y_train)
# Calculate R2 to see the accuracy of the model with train data.
r_sq_train4 = model4.score(X_train,y_train)
print('coefficient of determination with train data: %.3f' %r_sq_train4)

coefficient of determination with train data: 0.997


In [75]:
# Let's increase to change the solver to lbfgs (optimizer in the family of quasi-Newton methods)
model4 = MLPRegressor(activation='relu',solver='lbfgs',random_state=1, max_iter=500)
model4.fit(X_train,y_train)
# Calculate R2 to see the accuracy of the model with train data.
r_sq_train4 = model4.score(X_train,y_train)
print('coefficient of determination with train data: %.3f' %r_sq_train4)

        # Note: The default solver 'adam' works pretty well on relatively
        # large datasets (with thousands of training samples or more) in terms of
        # both training time and validation score.
        # For small datasets, however, 'lbfgs' can converge faster and perform
        # better.

coefficient of determination with train data: 1.000


#### Observation 9.
* Same as with Random Forest, R2 ise always 1, so we are not seen difference in accuracy changing the parameters.  
Probably when we take off the dependent variable these results will change.
* See comments in code, we have changed the number of hidden layers, affecting to the calculation speed of the model. 
* Also activation and solver was changed, but with no different results. (R2 = 1)


## Exercise 4 - Compare their performance using the test/train approach or using all data (internal validation).


# Level 2

## Exercise 5 Perform some variable engineering process to improve your prediction.



# Level 3 

## Exercise 6 - Do not use the DepDelay variable when making predictions.